src/java.base/share/classes/sun/security/pkcs/PKCS9Attribute.java

```diff
@@ -179,178 +179,126 @@ public class PKCS9Attribute implements DerEncoder {
+        ObjectIdentifier.of(KnownOIDs.ExtensionRequest);
     public static final ObjectIdentifier SIGNING_CERTIFICATE_OID =
-            PKCS9_OIDS[16] = ObjectIdentifier.of(KnownOIDs.SigningCertificate);
+        ObjectIdentifier.of(KnownOIDs.SigningCertificate);
     public static final ObjectIdentifier SIGNATURE_TIMESTAMP_TOKEN_OID =
-            PKCS9_OIDS[17] =
-            ObjectIdentifier.of(KnownOIDs.SignatureTimestampToken);
+        ObjectIdentifier.of(KnownOIDs.SignatureTimestampToken);
     public static final ObjectIdentifier CMS_ALGORITHM_PROTECTION_OID =
-            PKCS9_OIDS[18] =
-            ObjectIdentifier.of(KnownOIDs.CMSAlgorithmProtection);
+        ObjectIdentifier.of(KnownOIDs.CMSAlgorithmProtection);
 
-    /**
-     * Acceptable ASN.1 tags for DER encodings of values of PKCS9
-     * attributes, by index in <code>PKCS9_OIDS</code>.
-     * Sets of acceptable tags are represented as arrays.
-     */
-    private static final Byte[][] PKCS9_VALUE_TAGS = {
-        null,
-        {DerValue.tag_IA5String},   // EMailAddress
-        {DerValue.tag_IA5String,
-         DerValue.tag_PrintableString,
-         DerValue.tag_T61String,
-         DerValue.tag_BMPString,
-         DerValue.tag_UniversalString,
-         DerValue.tag_UTF8String},  // UnstructuredName
-        {DerValue.tag_ObjectId},    // ContentType
-        {DerValue.tag_OctetString}, // MessageDigest
-        {DerValue.tag_UtcTime,
-         DerValue.tag_GeneralizedTime}, // SigningTime
-        {DerValue.tag_Sequence},    // Countersignature
-        {DerValue.tag_PrintableString,
-         DerValue.tag_T61String,
-         DerValue.tag_BMPString,
-         DerValue.tag_UniversalString,
-         DerValue.tag_UTF8String},   // ChallengePassword
-        {DerValue.tag_PrintableString,
-         DerValue.tag_T61String,
-         DerValue.tag_BMPString,
-         DerValue.tag_UniversalString,
-         DerValue.tag_UTF8String},   // UnstructuredAddress
-        {DerValue.tag_SetOf},       // ExtendedCertificateAttributes
-        {DerValue.tag_Sequence},    // issuerAndSerialNumber
-        null,
-        null,
-        null,
-        {DerValue.tag_Sequence},    // extensionRequest
-        {DerValue.tag_Sequence},    // SMIMECapability
-        {DerValue.tag_Sequence},    // SigningCertificate
-        {DerValue.tag_Sequence},    // SignatureTimestampToken
-        {DerValue.tag_Sequence}     // CMSAlgorithmProtection
-    };
-
-    private static final Class<?>[] VALUE_CLASSES = new Class<?>[19];
+    private record AttributeInfo(byte[] valueTags, Class<?> valueClass, boolean singleValued) {}
+
+    private static final Map<ObjectIdentifier,AttributeInfo> oidMap = new HashMap<>();
+
+    private static void add(ObjectIdentifier oid, boolean singleValued,
+                            Class<?> valueClass, byte... valueTags) {
+        AttributeInfo info = new AttributeInfo(valueTags,valueClass,singleValued);
```

Comment:
Nit: add a space after ','.

Reply:
nit: match the ordering of arguments here w/ the one used in record?